src/Channels.Networking.TLS/SecureChannel.cs

```diff
@@ -0,0 +1,132 @@
+                            if (buff != null && buff.Length > 0)
+                            {
+                                var output = _lowerOutputChannel.Alloc(buff.Length);
+                                output.Write(buff);
+                                await output.FlushAsync();
+                            }
+                            if (securityContext.ReadyToSend)
+                            {
+                                StartWriting(securityContext);
+                            }
+                        }
+                        else if (f == TlsFrameType.Invalid)
+                        {
+                            throw new InvalidOperationException("We have recieved an invalid tls frame");
+                        }
+                        else if (f == TlsFrameType.AppData)
+                        {
+                            var messageBuffer = buffer.Slice(0, pointToSliceMessage);
+                            buffer = buffer.Slice(pointToSliceMessage);
+                            var decryptedData = _outputChannel.Alloc();
+
+                            securityContext.Decrypt(messageBuffer, decryptedData);
+
+                            await decryptedData.FlushAsync();
+                        }
+                        f = buffer.CheckForFrameType(out pointToSliceMessage);
+                    }
+                    _lowerInputChannel.Advance(buffer.Start, buffer.End);
+                }
+            }
+            finally
+            {
+                securityContext.Dispose();
+            }
+        }
+
+        private async void StartWriting<T>(T securityContext) where T : ISecureContext
+        {
+            int maxBlockSize = (SecurityContext.BlockSize - securityContext.HeaderSize - securityContext.TrailerSize);
+            while (true)
+            {
+                var buffer = await _inputChannel.ReadAsync();
+                while (buffer.Length > 0)
+                {
+                    ReadableBuffer messageBuffer;
+                    if (buffer.Length <= maxBlockSize)
+                    {
+                        messageBuffer = buffer;
+                        buffer = buffer.Slice(buffer.End);
+                    }
+                    else
+                    {
+                        messageBuffer = buffer.Slice(0, maxBlockSize);
+                        buffer = buffer.Slice(maxBlockSize);
+                    }
+
+                    var outputBuffer = _lowerOutputChannel.Alloc();
+
+                    securityContext.Encrypt(outputBuffer, messageBuffer);
+                    await outputBuffer.FlushAsync();
+                }
+                _inputChannel.Advance(buffer.End);
+            }
+        }
+
+        public void Dispose()
+        {
+            _inputChannel.CompleteReader();
+            _outputChannel.CompleteWriter();
+            _lowerOutputChannel.Complete();
```

Comment:
Instead of unwrapping the _lower* channels, why not just dispose the IChannel? That also possibly does more stuff.